File: node/api/src/db/AzureStorageTablesDatabase.js
const { TableClient, AzureNamedKeyCredential } = require('@azure/data-tables');
const DatabaseInterface = require('./DatabaseInterface');

class AzureStorageTablesDatabase extends DatabaseInterface {
  constructor(accountName, accountKey, tableName = 'urls') {
    super();
    this.tableName = tableName;
    this.credential = new AzureNamedKeyCredential(accountName, accountKey);
    this.client = new TableClient(
      `https://${accountName}.table.core.windows.net`,
      tableName,
      this.credential
    );
  }

  async initialize() {
    try {
      // Create table if it doesn't exist
      await this.client.createTable();
    } catch (err) {
      if (err.statusCode !== 409) { // 409 = table already exists
        throw new Error(`Azure Storage Tables initialization failed: ${err.message}`);
      }
    }
  }

  async createUrl(shortCode, originalUrl) {
    try {
      const entity = {
        partitionKey: 'urls',
        rowKey: shortCode,
        shortCode: shortCode,
        originalUrl: originalUrl,
        createdAt: new Date().toISOString()
      };

      await this.client.createEntity(entity);
      return { shortCode, originalUrl, id: shortCode };
    } catch (err) {
      throw new Error(`Failed to create URL: ${err.message}`);
    }
  }

  async getUrl(shortCode) {
    try {
      const entity = await this.client.getEntity('urls', shortCode);
      return {
        shortCode: entity.shortCode,
        originalUrl: entity.originalUrl,
        createdAt: entity.createdAt
      };
    } catch (err) {
      if (err.statusCode === 404) {
        return null;
      }
      throw new Error(`Failed to get URL: ${err.message}`);
    }
  }

  async getAllUrls() {
    try {
      const entities = [];
      const entitiesIter = this.client.listEntities({
        queryOptions: { filter: "PartitionKey eq 'urls'" }
      });

      for await (const entity of entitiesIter) {
        entities.push({
          shortCode: entity.shortCode,
          originalUrl: entity.originalUrl,
          createdAt: entity.createdAt
        });
      }

      // Sort by createdAt descending (client-side since Azure Tables doesn't support ORDER BY)
      return entities.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
    } catch (err) {
      throw new Error(`Failed to get all URLs: ${err.message}`);
    }
  }

  async urlExists(shortCode) {
    const url = await this.getUrl(shortCode);
    return !!url;
  }

  async close() {
    // Azure Tables client doesn't need explicit closing
  }
}

module.exports = AzureStorageTablesDatabase;

File: dotnet/UrlShortener.Core/Database/AzureStorageTablesDatabase.cs
using Azure;
using Azure.Data.Tables;
using UrlShortener.Core.Interfaces;
using UrlShortener.Core.Models;

namespace UrlShortener.Core.Database;

public class AzureStorageTablesDatabase : IUrlDatabase
{
    private readonly string _accountName;
    private readonly string _accountKey;
    private readonly string _tableName;
    private TableClient? _tableClient;

    public AzureStorageTablesDatabase(string accountName, string accountKey, string tableName = "urls")
    {
        _accountName = accountName;
        _accountKey = accountKey;
        _tableName = tableName;
    }

    public async Task InitializeAsync()
    {
        var serviceClient = new TableServiceClient(
            new Uri($"https://{_accountName}.table.core.windows.net"),
            new TableSharedKeyCredential(_accountName, _accountKey));

        _tableClient = serviceClient.GetTableClient(_tableName);
        await _tableClient.CreateIfNotExistsAsync();
    }

    public async Task<UrlMapping> CreateUrlAsync(string shortCode, string originalUrl)
    {
        if (_tableClient == null)
            throw new InvalidOperationException("Table client not initialized");

        var entity = new AzureTableEntity
        {
            PartitionKey = "urls",
            RowKey = shortCode,
            ShortCode = shortCode,
            OriginalUrl = originalUrl,
            CreatedAt = DateTime.UtcNow
        };

        await _tableClient.AddEntityAsync(entity);

        return new UrlMapping
        {
            ShortCode = entity.ShortCode,
            OriginalUrl = entity.OriginalUrl,
            CreatedAt = entity.CreatedAt
        };
    }

    public async Task<UrlMapping?> GetUrlAsync(string shortCode)
    {
        if (_tableClient == null)
            throw new InvalidOperationException("Table client not initialized");

        try
        {
            var response = await _tableClient.GetEntityAsync<AzureTableEntity>("urls", shortCode);
            var entity = response.Value;

            return new UrlMapping
            {
                ShortCode = entity.ShortCode,
                OriginalUrl = entity.OriginalUrl,
                CreatedAt = entity.CreatedAt
            };
        }
        catch (RequestFailedException ex) when (ex.Status == 404)
        {
            return null;
        }
    }

    public async Task<List<UrlMapping>> GetAllUrlsAsync()
    {
        if (_tableClient == null)
            throw new InvalidOperationException("Table client not initialized");

        var urls = new List<UrlMapping>();

        await foreach (var entity in _tableClient.QueryAsync<AzureTableEntity>(filter: "PartitionKey eq 'urls'"))
        {
            urls.Add(new UrlMapping
            {
                ShortCode = entity.ShortCode,
                OriginalUrl = entity.OriginalUrl,
                CreatedAt = entity.CreatedAt
            });
        }

        // Sort by created date descending (Azure Tables doesn't support ORDER BY)
        return urls.OrderByDescending(u => u.CreatedAt).ToList();
    }

    public async Task<bool> UrlExistsAsync(string shortCode)
    {
        var url = await GetUrlAsync(shortCode);
        return url != null;
    }

    public Task CloseAsync()
    {
        // TableClient doesn't require explicit cleanup
        return Task.CompletedTask;
    }

    private class AzureTableEntity : ITableEntity
    {
        public string PartitionKey { get; set; } = string.Empty;
        public string RowKey { get; set; } = string.Empty;
        public DateTimeOffset? Timestamp { get; set; }
        public ETag ETag { get; set; }

        public string ShortCode { get; set; } = string.Empty;
        public string OriginalUrl { get; set; } = string.Empty;
        public DateTime CreatedAt { get; set; }
    }
}
